File: php/divaserve.php
<?php

function get_max_zoom_level($img_wid, $img_hei, $t_wid, $t_hei) {
    $largest_dim = ($img_wid > $img_hei) ? $img_wid : $img_hei;
    $t_dim = ($img_wid > $img_hei) ? $t_wid : $t_hei;

    $zoom_levels = ceil(log(($largest_dim + 1)/($t_dim + 1), 2));
    return intval($zoom_levels);
}

function incorporate_zoom($img_dim, $zoom_difference) {
    return $img_dim / pow(2, $zoom_difference);
}

// Image directory. Images in this directory will be served by folder name,
// so if you have a folder named ...images/old_manuscript, a request for 
// divaserve.php?d=old_manuscript will use the images in that directory.
$IMAGE_DIR = "/mnt/images";
$CACHE_DIR = "/tmp/diva.js";
$IIP_SERVER = "http://localhost/cgi-bin/iipsrv.fcgi?";

### Nothing should change past here.
if (!isset($_GET['d']) || !isset($_GET['z'])) {
    die("Missing params");
}

$dir = preg_replace('/[^-a-zA-Z0-9_]/', '', $_GET['d']);
$zoom = preg_replace('/[^0-9]/', '', $_GET['z']);

$t_wid_get = (isset($_GET['t_w'])) ? $_GET['t_w'] : 0;
$t_hei_get = (isset($_GET['t_h'])) ? $_GET['t_h'] : 0;
$t_wid = (intval($t_wid_get) > 0) ? intval($t_wid_get) : 256;
$t_hei = (intval($t_hei_get) > 0) ? intval($t_hei_get) : 256;

// where we will store the text files.
$img_cache = $CACHE_DIR . "/" . $dir;
$img_dir = $IMAGE_DIR . "/" . $dir;

$cache_file = $img_cache . "/" . "docdata.txt";
$pgs = array();

if (!file_exists($img_cache)) {
    // Now go through the image directory and calculate stuff
    mkdir($img_cache);
    $images = array();
    $lowest_max_zoom = 0;
    foreach (glob($img_dir . '/*.tif') as $img_file) {
        $img_size = getimagesize($img_file);
        $img_wid = $img_size[0];
        $img_hei = $img_size[1];
        
        $max_zoom = get_max_zoom_level($img_wid, $img_hei, $t_wid, $t_hei);
        $lowest_max_zoom = ($lowest_max_zoom > $max_zoom || $lowest_max_zoom == 0) ? $max_zoom : $lowest_max_zoom;
        
        // Get the number from the filename (between the last _ and .)
        $img_num = intval(substr($img_file, strrpos($img_file, '_') + 1, strrpos($img_file, '.') - strrpos($img_file, '_') - 1));
        
        // Figure out the image filename
        $img_fn = substr($img_file, strrpos($img_file, '/') + 1);

        $images[$img_num] = array(
            'mx_h'      => $img_hei,
            'mx_w'      => $img_wid,
            'mx_z'      => $max_zoom,
            'fn'        => $img_fn,
        );
    }
    
    // Now go through them again, store in $pgs
    for ($i = 0; $i < count($images); $i++) {
        if (array_key_exists($i, $images)) {
            $h = incorporate_zoom($images[$i]['mx_h'], $lowest_max_zoom - $zoom);
            $w = incorporate_zoom($images[$i]['mx_w'], $lowest_max_zoom - $zoom);
            $c = ceil($w / $t_wid);
            $r = ceil($h / $t_hei);
            $m_z = $images[$i]['mx_z'];
            $fn = $images[$i]['fn'];
            $pgs[] = array(
                'c'     => $c,
                'r'     => $r,
                'h'     => $h,
                'w'     => $w,
                'm_z'   => $m_z,
                'fn'    => $fn, 
            );
        }
    }
} else {
    // Assume that the cache directory already contains everything we need
}
/*
// Calculate the dimensions
$dims = array(
    'a_wid'         => $a_wid,
    'a_hei'         => $a_hei,
    'mx_h'          => $mx_h,
    'mx_w'          => $mx_w,
    't_hei'         => $t_hei,
    't_wid'         => $t_wid
);*/

// The full data to be returned
$data = array(
    'item_title'    => $dir,
    'dims'          => '',//$dims,
    'max_zoom'      => $lowest_max_zoom,
    'pgs'           => $pgs
);

echo json_encode($data);

/*
if($handle = opendir($img_dir)) {
    $img_info = array();
    $file_num = 0;
    $max_res_number = 0;
    
    while (false !== ($file = readdir($handle))) {
        
        if ($filenum > 10) {
           break;
        }
        
        if($file == "." && $file =="..") {
            continue;
        }
        
        $fullpath = $imgdir . "/" . $file;
        $pthpts = pathinfo($fullpath);
        $extn = $pthpts['extension'];
        $name = $pthpts['filename'];

        // echo $fullpath;

        if (($extn != "tif") && ($extn != "tiff")) {
            continue;
        }
        
        $filenum++;

        # check if a cached file of the image info exists for this image. If not, make one.
        $tfile = $imgcache . "/" . $name . ".txt";
        if(!file_exists($tfile)) {
            $image_array = array();
             
            $imginfo = file_get_contents($IIP_SERVER . "FIF=" . $fullpath . "&obj=IIP,1.0&obj=Max-size&obj=Tile-size&obj=Resolution-number", 'r');
            $t_array = split("\r\n", $imginfo);

            $mxs = split(":", $t_array[1]);
            $max_size = split(" ", $mxs[1]);

            $tls = split(":", $t_array[2]);
            $tile_size = split(" ", $tls[1]);

            $rsn = split(":", $t_array[3]);
            $res_number = (int) $rsn[1];
            
            if ($res_number > $max_res_number) {
                $max_res_number = $res_number;
            }
            
            $w = $max_size[0];
            $h = $max_size[1];
            $t_w = $tile_size[0];
            $t_h = $tile_size[1];

            $image_array = array(
                "w" => $w,
                "h" => $h,
                "r" => $max_res_number,
                "t_w" => $t_w,
                "t_h" => $t_h,
                "fn" => $fid . "/" . $file
                );
                
            array_push( $imginfos, $image_array );
        }
     }
        
     //print_r($res_number);
     for ($i=0; $i < $max_res_number; $i++) {
        
        $pgs = array();
        $max_width = 0;
        $max_height = 0;
        $avg_width = 0;
        $avg_height = 0;
        $total_width = 0;
        $total_height = 0;
        
        for($j=0; $j < $filenum; $j++) {
           
           $cur_res = $i;
           if ($imginfos[$j]["r"] - 1 < $i) {
              $cur_res = $imginfos[$j]["r"] - 1;
           }
           
           $w = $imginfos[$j]["w"];
           $h = $imginfos[$j]["h"];
           $t_w = $imginfos[$j]["t_w"];
           $t_h = $imginfos[$j]["t_h"];
           $zwid = $w / pow( 2, $imginfos[$j]["r"] - $cur_res - 2 );
           $zhei = $h / pow( 2, $imginfos[$j]["r"] - $cur_res - 2 );
           $row = ceil( $zhei / $t_h );
           $col = ceil( $zwid / $t_w );
           $tot = ( $row * $col );
           
           $imgresp = array(
               "w" => $zwid,
               "h" => $zhei,
               "z" => $i,
               "r" => $row,
               "c" => $col,
               "t_til" => $tot,
               "t_wid" => $t_w,
               "t_hei" => $t_h,
               "fn" => $imginfos[$j]["fn"]
               );
            
            array_push( $pgs, $imgresp );
            
           // update some global variables for this collection.
           if ($zwid > $max_width) {
               $max_width = $zwid;
           }
           if ($zhei > $max_height) {
               $max_height = $zhei;
           }
           $total_width += $zwid;
           $total_height += $zhei;
           
        }
        
        $dims = array(
            "mx_w" => $max_width,
            "mx_h" => $max_height,
            "t_wid" => $total_width,
            "t_hei" => $total_height,
            "a_wid" => $avg_width,
            "a_hei" => $avg_height
            );
            
         if ($zlv == $i) {
            break;
         }

      }
};

$resp = array(
    'dims' => $dims,
    'pgs' => $pgs
    //"item_title" => $fid
);

echo json_encode($resp);

exit;


$pgs = array();
if($handle = opendir($imgcache)) {
    while (false !== ($file = readdir($handle))) {
        if($file == "." || $file ==".." || $file == "docdata.txt") {
            continue;
        }
        $fullpath = $imgcache . "/" . $file;
        $pthpts = pathinfo($fullpath);
        $extn = $pthpts['extension'];
        $name = $pthpts['filename'];
        
        $imginfo = file_get_contents($fullpath);
        $f_data = unserialize($imginfo);
        $f_zdata = $f_data[$zlv]; // get the file's data at a given zoom level
        $pgs[] = $f_zdata;
    } //end while
} // end open handle
$f_dims = file_get_contents($gfile);
$dims = unserialize($f_dims);

$resp = array(
    'dims' => $dims,
    'pgs' => $pgs,
    "item_title" => $fid
);

echo json_encode($resp);
*/
?>

File: python/divaserve.py
import os
import sys
import math
try:
    from vipsCC import VImage
    def image_size(fn):
        img = VImage.VImage(fn)
        size = (img.Xsize(), img.Ysize())
        del img
        return size
except ImportError:
    try:
        from PIL import Image
    except ImportError:
        import Image
    def image_size(fn):
        print fn
        img = Image.open(fn)
        size = img.size
        del img
        return size
from operator import itemgetter

class DivaServe(object):
    def __init__(self, directory, t=256, mode="memory"):
        """ <directory> is the directory of images to serve 
            "mode" can be either "memory" or "disk" for the cache mode to use
            if "mode" is "disk," you must also set "cachedir".
            <t> is be the tilesize. Default 256
        """

        self.imgdir = directory
        self.images = {}
        self.data = {}
        self.lowest_max_zoom = 0

        self.til_wid = t
        self.til_hei = t

        lmz = 0
        # Read image sizes on startup
        print >> sys.stderr, "loading images..."
        for i,f in enumerate(os.listdir(self.imgdir)):
            if os.path.splitext(f)[1] not in (".tif", ".tiff"):
                continue
            img_wid, img_hei = image_size(os.path.join(self.imgdir, f))

            max_zoom = self._get_max_zoom_level(img_wid, img_hei, self.til_wid, self.til_hei)
            if max_zoom > lmz:
                lmz = max_zoom
 
            self.images[i] = {
                'mx_w': img_wid,
                'mx_h': img_hei,
                'mx_z': max_zoom,
                'fn': f
            }

        self.lowest_max_zoom = lmz
        print >> sys.stderr, "images loaded"


    def get(self, zoom):
        if zoom in self.data.keys() and len(self.data[zoom]) > 0:
            return self.data[zoom]

        if zoom > self.lowest_max_zoom:
            zoom = 0
        elif zoom < 0:
            zoom = 0

        mx_h = mx_w = t_wid = t_hei = num_pages = max_ratio = 0
        pgs = []

        for v in self.images.itervalues():
            h = self._incorporate_zoom(v['mx_h'], self.lowest_max_zoom - zoom)
            w = self._incorporate_zoom(v['mx_w'], self.lowest_max_zoom - zoom)

            c = math.ceil(w / float(self.til_wid))
            r = math.ceil(h / float(self.til_hei))
            m_z = v['mx_z']
            fn = v['fn']
            
            pgs.append({
                'c': c,
                'r': r,
                'h': h,
                'w': w,
                'm_z': m_z,
                'fn': fn
            })

            if h > mx_h:
                mx_h = h

            if w > mx_w:
                mx_w = w
            
            ratio = h / float(w)
            max_ratio = ratio if ratio > max_ratio else max_ratio
            t_wid += w
            t_hei += h
            num_pages += 1
        pgs.sort(key=itemgetter("fn"))

        if num_pages > 0:
            a_wid = t_wid / float(num_pages)
            a_hei = t_hei / float(num_pages)

            dims = {
                'a_wid': a_wid,
                'a_hei': a_hei,
                'max_w': mx_w,
                'max_h': mx_h,
                'max_ratio': max_ratio,
                't_hei': t_hei,
                't_wid': t_wid
            }

            title = os.path.basename(self.imgdir).replace('-', ' ').title()

            self.data[zoom] = {
                'item_title': title,
                'dims': dims,
                'max_zoom': self.lowest_max_zoom,
                'pgs': pgs
            }
        else:
            self.data[zoom] = {}

        return self.data[zoom]

    def _get_max_zoom_level(self, iwid, ihei, twid, thei):
        largest_dim = max(iwid, ihei)
        t_dim = twid if iwid > ihei else thei

        zoom_levels = math.ceil(math.log((largest_dim + 1) / float(t_dim) + 1, 2))
        return int(zoom_levels)

    def _incorporate_zoom(self, img_dim, zoom_diff):
        return img_dim / float(2**zoom_diff)

class DivaException(BaseException):
    def __init__(self, message):
        self.message = message
    def __str__(self):
        return repr(self.message)
